lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,302 @@
+      #Nx.Tensor<
+        f32[4][2]
+        [
+          [11.721418380737305, 0.0],
+          [74.18099975585938, 0.0],
+          [-24.942012786865234, 0.0],
+          [-60.960227966308594, 0.0]
+        ]
+      >
+  """
+  deftransform fit(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(x, seed, opts)
+  end
+
+  defnp fit_n(x, seed, opts \\ []) do
+    {perplexity, learning_rate, num_iters, num_components, exaggeration, init, metric} =
+      {opts[:perplexity], opts[:learning_rate], opts[:num_iters], opts[:num_components],
+       opts[:exaggeration], opts[:init], opts[:metric]}
+
+    x = to_float(x)
+    {n, _dims} = Nx.shape(x)
+
+    y1 =
+      case init do
+        :random ->
+          key = Nx.Random.key(seed)
+
+          {y, _new_key} =
+            Nx.Random.normal(key, 0.0, 1.0e-4, shape: {n, num_components}, type: Nx.type(x))
+
+          y
+
+        :pca ->
+          Scholar.Decomposition.PCA.fit_transform(x, num_components: num_components)
+      end
+
+    p = p_joint(x, perplexity, metric)
+
+    {y, _, _, _, _} =
+      while {y1, y2 = y1, learning_rate, p, i = 2}, i < num_iters do
+        q = q_joint(y1, metric)
+        grad = gradient(p * exaggeration(i, exaggeration), q, y1, metric)
+        y_next = y1 - learning_rate * grad + momentum(i) * (y1 - y2)
+        {y_next, y1, learning_rate, p, i + 1}
+      end
+
+    y
+  end
+
+  defnp pairwise_dist(x, metric) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    t1 =
+      x
+      |> Nx.reshape({1, num_samples, num_features})
+      |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    t2 =
+      x
+      |> Nx.reshape({num_samples, 1, num_features})
+      |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    case metric do
+      :squared_euclidean ->
+        Distance.squared_euclidean(t1, t2, axes: [2])
+
+      :euclidean ->
+        Distance.euclidean(t1, t2, axes: [2])
+
+      :manhattan ->
+        Distance.manhattan(t1, t2, axes: [2])
+
+      :cosine ->
+        Distance.cosine(t1, t2, axes: [2])
+
+      :chebyshev ->
+        Distance.chebyshev(t1, t2, axes: [2])
+    end
+  end
+
+  defnp p_conditional(distances, sigmas) do
+    arg = -distances / (2 * Nx.reshape(sigmas, {:auto, 1})) ** 2
+    {n, _} = Nx.shape(arg)
+
+    # Set diagonal to a large negative number so it becomes 0 after applying exp()
+    min_value = Nx.Constants.min_finite(Nx.type(arg))
+    arg_with_min_diag = Nx.put_diagonal(arg, Nx.broadcast(min_value, {n}))
+
+    stabilization_constant = Nx.reduce_max(arg_with_min_diag, axes: [1], keep_axes: true)
+    arg = arg - stabilization_constant
+
+    p = Nx.exp(arg)
+    p / Nx.sum(p, axes: [1], keep_axes: true)
+  end
+
+  defnp perplexity(condition_matrix) do
+    exponent = -Nx.sum(condition_matrix * Nx.log2(condition_matrix), axes: [1])
+    2 ** exponent
+  end
```

Comment:
```suggestion
  defnp perplexity(p_matrix) do
    # Nx.select is used below so that if the entry is eps or less, we treat it as 0,
    # and this makes it so we can avoid 0 * -inf == nan issues 
    eps = Nx.Constants.eps(Nx.type(p_matrix)
    shannon_entropy_partials = Nx.select(p_matrix <= eps, 0, p_matrix * Nx.log2(p_matrix))
    shannon_entropy = -Nx.sum(h_partials, axes: [1])
    2 ** shannon_entropy
  end
```

There's a possibly neat optimization in terms of numerical stability that we could also apply, but it's slightly overkill:

```elixir
p = 
  case p_matrix do
    %{vectorized_axes: []} -> Nx.vectorize(p, :rows)
    %{vectorized_axes: [{first_axis, _} | _]} -> Nx.revectorize(p, [{first_axis, :auto}])
  end
  
eps = Nx.Constants.eps(Nx.type(p_matrix)

perplexity = 
  cond do
    Nx.all_close(p, p[0]) ->
      # if all elements of p are equal, then the entropy is log2(size(p))
      # which means that the perplexity is 2 ** log2(size(p)) === size(p)
      Nx.size(n)
    
    Nx.reduce_max(p) >= 1 - eps ->
      # if the maximum value is approximately 1, it means that all terms
      # of the sum will be approximately 0, which leads to 2 ** 0 = 1
      1
      
    true ->
      # Nx.select is used below so that if the entry is eps or less, we treat it as 0,
      # and this makes it so we can avoid 0 * -inf == nan issues
      shannon_entropy_partials = Nx.select(p_matrix <= eps, 0, p_matrix * Nx.log2(p_matrix))
      shannon_entropy = -Nx.sum(h_partials)
      2 ** shannon_entropy
  end
  
Nx.revectorize(perplexity, p.vectorized_axes)
```

IMO, first version is sufficient, but second one accounts for more edge cases using [_math_](https://xkcd.com/1131/)

Reply:
I think I will stay with the first version for now, WDYT @josevalim?